src/panel/plugin/plugin-manager.c

```diff
@@ -0,0 +1,472 @@
+ * @self: A #BudgiePanelPluginManager instance.
+ * @name: The name of a plugin.
+ *
+ * Checks if the plugin with name @name is valid. We do this
+ * by trying to get the #PeasPluginInfo with @name, and returning
+ * %TRUE if it is found.
+ *
+ * Returns: %TRUE if the plugin is valid, %FALSE otherwise.
+ */
+gboolean budgie_panel_plugin_manager_is_plugin_valid(BudgiePanelPluginManager *self, const gchar *name) {
+	g_return_val_if_fail(BUDGIE_IS_PANEL_PLUGIN_MANAGER(self), FALSE);
+	g_return_val_if_fail(name != NULL, FALSE);
+
+	g_autoptr(PeasPluginInfo) info = NULL;
+
+	info = get_plugin_info(self, name);
+
+	return PEAS_IS_PLUGIN_INFO(info);
+}
+
+/**
+ * budgie_panel_plugin_manager_rescan_plugins:
+ * @self: A #BudgiePanelPluginManager instance.
+ *
+ * Triggers a re-scan by the #PeasEngine for plugins.
+ */
+void budgie_panel_plugin_manager_rescan_plugins(BudgiePanelPluginManager *self) {
+	g_return_if_fail(BUDGIE_IS_PANEL_PLUGIN_MANAGER(self));
+
+	peas_engine_garbage_collect(self->engine);
+	peas_engine_rescan_plugins(self->engine);
+}
+
+/**
+ * budgie_panel_plugin_manager_modprobe:
+ * @self: A #BudgiePanelPluginManager instance.
+ * @name: The name of a plugin.
+ *
+ * Tells the #PeasEngine to load the plugin with the name @name.
+ */
+void budgie_panel_plugin_manager_modprobe(BudgiePanelPluginManager *self, const gchar *name) {
+	g_return_if_fail(BUDGIE_IS_PANEL_PLUGIN_MANAGER(self));
+	g_return_if_fail(name != NULL);
+
+	g_autoptr(PeasPluginInfo) info = NULL;
+	gboolean success = FALSE;
+
+	info = get_plugin_info(self, name);
+
+	if G_UNLIKELY (!PEAS_IS_PLUGIN_INFO(info)) {
+		g_warning("modprobe called for non existent module: %s", name);
+		return;
+	}
+
+	success = peas_engine_load_plugin(self->engine, info);
+
+	if (!success) {
+		g_warning("Failed to load plugin with name '%s'", name);
+	}
+}
+
+/**
+ * budgie_panel_plugin_manager_load_applet_instance:
+ * @self: A #BudgiePanelPluginManager instance.
+ * @uuid: A plugin's UUID.
+ * @settings: (optional): The plugin's #GSettings instance.
+ * @name: (out callee-allocates): (optional): A return location for a plugin's name.
+ * @err: (out callee-allocates): (optional): A return location for a #GError.
+ *
+ * Attempt to load an instance of a Budgie plugin. If the plugin
+ * could not be loaded, %NULL will be returned, and @err will be set.
+ *
+ * Returns: (transfer full): The plugin's #BudgieAppletInfo if loaded, or %NULL.
+ */
+BudgieAppletInfo *budgie_panel_plugin_manager_load_applet_instance(BudgiePanelPluginManager *self, const gchar *uuid, GSettings *plugin_settings, gchar **name, GError **err) {
+	g_return_val_if_fail(BUDGIE_IS_PANEL_PLUGIN_MANAGER(self), NULL);
+	g_return_val_if_fail(uuid != NULL, NULL);
+	g_return_val_if_fail(name == NULL || *name == NULL, NULL);
+	g_return_val_if_fail(err == NULL || *err == NULL, NULL);
+
+	g_autofree gchar *path = NULL;
+	GSettings *settings = NULL;
+	g_autofree gchar *plugin_name = NULL;
+	PeasPluginInfo *info = NULL;
```

Comment:
The old API says that the caller owns the instance - so shouldn't we cleanup usage of info?

Reply:
I have to double check the context, but I'm pretty sure the info we get is owned by Peas, so we dup it when we return it.